---
title: "Vervet_UCP1"
author: "Christian Gagnon"
date: "1/6/2019"
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown for Chlorocebus pygerythrus

###Coordinates
Human UCP1 coordinates 
chr4:140,556,520-140,565,891
Human UCP1 Coordinates including promoter region:
chr4:140,554,000-140,570,000
Vervet UCP1 coordinates including promoter
CAE7:87,486,745-87,503,719

###Unix code for obtaining UCP1, UCP2 and UCP3 regions. 

We are going to look at three gene homologs: UCP1, UCP2, and UCP3. First, we need to isolate each of the three genes separately from the vervet genome data. Then, we will isolate them for the three taxa of the Southern Expansion populations: Cynosuros, Hilgerti, and Pygerythrus. We will also isolate the genes separately for cynosuros and pygerythrus for finding neutrality statistics to look for selection (we are not doing this for hilgerti because there are too few samples). Lastly, we will get a list of variants from each isolated gene to use for our linkage disequillibrium tests. We're going to do this all using the tabix module on the SCC: 
```{bash}
module load tabix
module load vcftools

#UCP1
tabix -h  163_201701_UG_vervet_monkey_SNPs_all_chrom_beagle_shapeit.vcf.gz CAE7:87486745-87503719 | bgzip -c > ALL_UPC1.vcf.gz #isolate UCP1 region for all samples

vcf-subset -c southernexpansion_list ALL_UPC1.vcf.gz | bgzip  -c > se_UCP1.vcf.gz #isolate only southern expansion samples

vcf-subset -c cynosuros.txt ALL_UPC1.vcf.gz | bgzip  -c > cyn_UCP1.vcf.gz #isolate cynosuros samples for neutrality stats

vcf-subset -c pygerythrus.txt ALL_UPC1.vcf.gz | bgzip  -c > pyg_UCP1.vcf.gz #isolate pygerythrus samples for neutrality stats

zgrep -v "^##" se_UCP1.vcf.gz | cut -f1-3 > UCP1_loci.txt #Get list of variant sites for LD

#UCP2
tabix -h  163_201701_UG_vervet_monkey_SNPs_all_chrom_beagle_shapeit.vcf.gz CAE1:65228307-65256886 | bgzip -c > ALL_UPC2.vcf.gz #isolate UCP2 region for all samples

vcf-subset -c southernexpansion_list ALL_UPC2.vcf.gz | bgzip  -c > se_UCP2.vcf.gz #isolate only southern expansion samples

vcf-subset -c cynosuros.txt ALL_UPC2.vcf.gz | bgzip  -c > cyn_UCP2.vcf.gz #isolate cynosuros samples for neutrality stats

vcf-subset -c pygerythrus.txt ALL_UPC2.vcf.gz | bgzip  -c > pyg_UCP2.vcf.gz #isolate pygerythrus samples for neutrality stats

zgrep -v "^##" se_UCP2.vcf.gz | cut -f1-3 > UCP2_loci.txt #Get list of variant sites for LD

#UCP3
tabix -h  163_201701_UG_vervet_monkey_SNPs_all_chrom_beagle_shapeit.vcf.gz CAE2:65259611-65289050 | bgzip -c > ALL_UPC3.vcf.gz #isolate UCP3 region for all samples

vcf-subset -c southernexpansion_list ALL_UPC3.vcf.gz | bgzip  -c > se_UCP3.vcf.gz #isolate only southern expansion samples

vcf-subset -c cynosuros.txt ALL_UPC3.vcf.gz | bgzip  -c > cyn_UCP3.vcf.gz #isolate cynosuros samples for neutrality stats

vcf-subset -c pygerythrus.txt ALL_UPC3.vcf.gz | bgzip  -c > pyg_UCP3.vcf.gz #isolate pygerythrus samples for neutrality stats

zgrep -v "^##" se_UCP3.vcf.gz | cut -f1-3 > UCP3_loci.txt #Get list of variant sites for LD
```

###Hardy-Weinberg Equilibrium
Let's see if these variants are in HWE. To do so, first we can go to our *R* space and run a quick power test to see if 51 samples is even enough to see anything:

chi-squared test to check if sample size is adequate 
```{r,eval=F,echo=T}
library(pwr)
pwr.chisq.test(w = 0.5, df = 1, sig.level = 0.05, power=0.95)
```

Hmmm... the above equation is for 52 samples (we have 51). After playing around a bit, it looks like we'd really need a pretty big effect size (0.5) to see a statistically strong result given our sample size.

Let's take a look!

```{bash, eval = F, echo = T, class.source="sccCode"}
module load vcftools
vcftools --vcf se_UCP3.vcf --hardy --out se_UCP3
less se_UCP3.hwe

```

Analysis:
```{r}
library(vcfR)
library(pegas)
#UCP1
UCP1 <- read.vcfR("~/Desktop/final/se_UCP1.vcf.gz", verbose = TRUE) #Read in data
pegas.ALL <- vcfR2genind(ALL, sep = "[|/]") #Use pegas to perform analysis
HWE <- hw.test(pegas.ALL, B = 0) #copy output to new file called HWE
HWE #View output

#UCP2
ALL <- read.vcfR("~/Desktop/final/se_UCP2.vcf.gz", verbose = TRUE) 
pegas.ALL <- vcfR2genind(ALL, sep = "[|/]") 
HWE <- hw.test(pegas.ALL, B = 0) 
HWE #View output
#UCP3
ALL <- read.vcfR("~/Desktop/final/se_UCP2.vcf.gz", verbose = TRUE) #Read in data
pegas.ALL <- vcfR2genind(ALL, sep = "[|/]") #Use pegas to perform analysis
HWE <- hw.test(pegas.ALL, B = 0) #copy output to new file called HWE
HWE #View output

```
Convert the vcf of UCP1 data into a GENIND object.

```{r}
UCP1 <- read.vcfR("~/Desktop/final/se_UCP1.vcf.gz", verbose = TRUE) #Read in data
UCP1_genind <- vcfR2genind(UCP1, sep = "[|/]") #convert to GENIND
```

Now we have our object, but it doesn't have the population info assigned to each sample. This time, we are breaking up the Free State region's population into north and south. The simplest way to add population parameters to the 'vcf' R file is to run this code on your GENIND object (which has a slot for population information):
```{r}
library(adegenet)
pop(UCP1_genind)<-as.factor(c("cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","FS North","FS North","FS North","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS North","FS North","FS North","FS North","FS North","FS North","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","VSAI3005","FS North","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","hilgerti","hilgerti","hilgerti","hilgerti","cynosuros","cynosuros","hilgerti","hilgerti"))

fstat(UCP1_genind) # calculating Fst
```

This gives us Fst all of the populations combined. Now let's look at the pairwise Fst between each population: 
```{r}
UCP1Fst <- pairwise.fst(UCP1_genind[1:73,])
UCP2Fst
```

Now let's do it for the other UCPs:

UCP2:
```{r}
UCP2 <- read.vcfR("~/Desktop/final/se_UCP2.vcf.gz", verbose = TRUE) #Read in data
UCP2_genind <- vcfR2genind(UCP2, sep = "[|/]") #convert to GENIND
pop(UCP2_genind)<-as.factor(c("cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","FS North","FS North","FS North","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS North","FS North","FS North","FS North","FS North","FS North","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","VSAI3005","FS North","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","hilgerti","hilgerti","hilgerti","hilgerti","cynosuros","cynosuros","hilgerti","hilgerti"))
fstat(UCP2_genind) # overall Fst
UCP2Fst <- pairwise.fst(UCP2_genind[1:73,]) # calculate pairwise values
UCP2Fst
```
UCP3:
```{r}
UCP3 <- read.vcfR("~/Desktop/final/se_UCP3.vcf.gz", verbose = TRUE) #Read in data
UCP3_genind <- vcfR2genind(UCP3, sep = "[|/]") #convert to GENIND
pop(UCP3_genind)<-as.factor(c("cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","cynosuros","FS North","FS North","FS North","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS South","FS North","FS North","FS North","FS North","FS North","FS North","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","KZN","VSAI3005","FS North","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","Eastern Cape","hilgerti","hilgerti","hilgerti","hilgerti","cynosuros","cynosuros","hilgerti","hilgerti"))
fstat(UCP3_genind) # overall Fst
UCP3Fst <- pairwise.fst(UCP3_genind[1:73,]) # calculate pairwise values
UCP3Fst
```


Visualize Fst: 
```{r}
# install.packages("qqman")
# library(qqman)

# fst <- read.table("~/Desktop/final/UCP1F.weir.fst", header=TRUE)
# fstsubset<-fst[complete.cases(fst),]
# SNP<-c(1:(nrow(fstsubset)))
# mydf<-data.frame(SNP,fstsubset)
```

###Neutrality statistics

First, we will look at Tajima's D. This tests whether mutations in the UCP regions in a population are following or violating the assumptions of neutrality models. If they are not, this implies selection is acting on them. When D is close to zero, there are no alleles over or under represented, meaning they are neutral. If D is negative, this implies purifying selection, while if D is positive, the population is experiencing positive selection on the region. 

Trying to do it in R???
```{r}
library(PopGenome)
library(vcfR)
cynUCP1 <- read.vcfR("~/Desktop/final/cyn_UCP2.vcf.gz")
cyngenome <- readData("~/Desktop/final/cyn_UCP2.vcf.gz", format = "VCF")
neut <- neutrality.stats(cynUCP1)
```

Using vcftools: 
```{bash}
# Cynosuros
#UCP1
vcftools --gzvcf cyn_UCP1.vcf.gz --TajimaD 500 # We chose 500 as the window size 
mv out.Tajima.D cynUCP1.Tajima.D #rename the output file 
mv out.log cynUCP1TD.log #rename log file 
#UCP2
vcftools --gzvcf cyn_UCP2.vcf.gz --TajimaD 500
mv out.Tajima.D cynUCP2.Tajima.D
mv out.log cynUCP2TD.log
#UCP3
vcftools --gzvcf cyn_UCP3.vcf.gz --TajimaD 500
mv out.Tajima.D UCP3.Tajima.D
mv out.log UCP3D.log
```

FU and Li's D and F statistic are a little similar in that anything deviating from 0 is a violation of neutrality. The D statistic is based on the difference between the number of private mutations in the sample and the total number of mutations. The F statistic is based on the difference between private mutations and average number of nucleotide differences between sequence pairs. 

Packages:
```{r}
library(vcfR)
library(PopGenome)
```


###Linkage Disequilibrium

Packages:
```{r}
install.packages("LDheatmap")
install.packages("VariantAnnotation")
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("snpStats", version = "3.8")
#install.packages("snpStats")
```


Analysis:
```{r}
library(snpStats)
library(VariantAnnotation)
library(LDheatmap)

vcfbig <- readVcf("se_UCP3.vcf.gz", "hg19")
ALLmatrix <- genotypeToSnpMatrix(vcfbig)
ALLmatrix

LD <- ld(ALLmatrix$genotypes, depth = 200, stats = "R.squared")
head(LD)
LD
```

Heat Mapper:
```{r}
cols = colorRampPalette(c("yellow", "red"))(10)
image(LD, lwd = 0, cuts= 9, col.regions=cols, colorkey=TRUE)
positions <- c(87492144, 87492145, 87492201, 87492261, 87492339, 
               87492384, 87492403, 87492424, 87492537, 87492563, 
               87492569, 87492579, 87492612, 87492667, 87492739, 
               87492801, 87492842, 87492898, 87492907, 87492922, 
               87492949, 87492998, 87493023, 87493031, 87493125, 
               87493148, 87493153, 87493215, 87493237, 87493263, 
               87493283, 87493291, 87493379, 87493469, 87493472, 
               87493519, 87493569, 87493613, 87493664, 87493744, 
               87493753, 87493830, 87493837, 87493854, 87493882, 
               87493885, 87493888, 87493967, 87493997, 87494045, 
               87494126, 87494157, 87494177, 87494188, 87494216, 
               87494226, 87494338, 87494339, 87494341, 87494342, 
               87494406, 87494423, 87494503, 87494508, 87494576, 
               87494632, 87494638, 87494674, 87494842, 87494859, 
               87494861, 87494916, 87495040, 87495094, 87495117, 
               87495209, 87495213, 87495244, 87495288, 87495371, 
               87495400, 87495483, 87495511, 87495557, 87495566, 
               87495638, 87495673, 87495675, 87495720, 87495756, 
               87495918, 87496052, 87496059, 87496184, 87496254, 
               87496359, 87496412, 87496425, 87496460, 87496622, 
               87496650, 87496753, 87496822, 87496860, 87496865, 
               87496877, 87496889, 87496897, 87496950, 87496956, 
               87496971, 87497024, 87497044, 87497047, 87497115, 
               87497190, 87497200, 87497259, 87497358, 87497419, 
               87497423, 87497427, 87497467, 87497483, 87497539, 
               87497546, 87497583, 87497753, 87497809, 87497815, 
               87497848, 87497910, 87497920, 87498023, 87498068, 
               87498167, 87498192, 87498326, 87498342, 87498397, 
               87498450, 87498470, 87498487, 87498501, 87498511, 
               87498532, 87498564, 87498575, 87498578, 87498611, 
               87498614, 87498628, 87498633, 87498652, 87498662, 
               87498678, 87498681, 87498695, 87498769, 87498903, 
               87498904, 87498981, 87499037, 87499064, 87499099, 
               87499115, 87499116, 87499128, 87499152, 87499154, 
               87499168, 87499175, 87499179, 87499180, 87499205, 
               87499292, 87499295, 87499303, 87499326, 87499356, 
               87499382, 87499434, 87499439, 87499443, 87499459, 
               87499485, 87499618, 87499641, 87499761, 87499800, 
               87499801, 87499818, 87499828, 87499840, 87499854, 
               87499902, 87500036, 87500045, 87500110, 87500111, 
               87500152, 87500233, 87500305, 87500306, 87500313, 
               87500323, 87500325, 87500408, 87500413, 87500454, 
               87500458, 87500545, 87500558, 87500605, 87500606, 
               87500611, 87500623, 87500632, 87500650, 87500651, 
               87500686, 87500780, 87500821, 87500871, 87500927, 
               87500934, 87500935, 87500951, 87500984, 87501102, 
               87501183, 87501221, 87501233, 87501339, 87501420, 
               87501475, 87501486, 87501513, 87501527, 87501535, 
               87501560, 87501581, 87501599, 87501638, 87501662, 
               87501699, 87501756, 87501805, 87501834, 87501898, 
               87501903, 87501904, 87501910, 87501922, 87501946, 
               87501948, 87501955, 87501974, 87501983, 87502069, 
               87502083, 87502085, 87502127, 87502141, 87502149, 
               87502154, 87502235, 87502245, 87502246, 87502255, 
               87502264, 87502276, 87502281, 87502305, 87502325, 
               87502347, 87502374, 87502418, 87502422, 87502428, 
               87502449, 87502469, 87502477, 87502481, 87502522)

allheatmapr <- LDheatmap(ALLmatrix$genotypes,
                           genetic.distances=positions,
                           distances="physical",
                           LDmeasure="r",
                           title="Pairwise LD with R^2",
                           add.map=TRUE, add.key=TRUE,
                           geneMapLocation=0.15,
                           SNP.name=c("87493148"),
                           color=NULL, newpage=TRUE,
                           name="ldheatmap")

allheatmapD <- LDheatmap(ALLmatrix$genotype,
                           genetic.distances=positions,
                           distances="physical",
                           LDmeasure="D",
                           title="Pairwise LD with D'",
                           add.map=TRUE, add.key=TRUE,
                           geneMapLocation=0.15,
                           SNP.name=c("87493148"),
                           color=NULL, newpage=TRUE,
                           name="ldheatmap")
View(allheatmapr$LDmatrix)
View(allheatmapD$LDmatrix)
```

###Phylogenetic Analysis

Packages:
```{r}
library(vcfR)
library(ape)
library(phangorn)
library(ade4)
```


Read in data:(finding Fst 4.23.19)
```{r}
VER <- read.vcfR("se_UCP1.vcf.gz", verbose = TRUE, nrows=285)

VERdna <- vcfR2DNAbin(VER)
VERdna
dist <- dist.dna(VERdna, model = "K80")
length(dist)
heatmap <- as.data.frame(as.matrix(dist))
table.paint(heatmap, cleg=0, clabel.row=.5, clabel.col=.5)
VERtree <- nj(dist)
class(VERtree) #tree should be class "phylo"
```

#Just a quick sneak peek at our tree before we plot it
```{r}
summary.phylo(VERtree)
plot(VERtree, cex=0.5)
```


What to do here?
```{r}
#Tree 1
#CLM <- read.vcfR("CLM.vcf", verbose = TRUE, nrows=206)
#CLMdna <- vcfR2DNAbin(CLM)
#CLMdna
#D <- dist.dna(CLMdna, model = "K80")
#length(D)
#CLMtree <- nj(D)
#class(CLMtree)
#Tree 2
#MXL <- read.vcfR("MXL.vcf", verbose = TRUE, nrows=206)
#MXLdna <- vcfR2DNAbin(MXL)
#MXLdna
#D <- dist.dna(MXLdna, model = "K80")
#length(D)
#MXLtree <- nj(D)
#class(MXLtree)
#samerica <- c(PELtree, CLMtree, MXLtree)
#class(samerica)

#densitree <- densiTree(samerica, type="phylogram", col=c("red", "green", "blue"), width=2, jitter=list(amount=.3, random=FALSE), alpha=1)
```

##Statistics

Packages:
```{r}
install.packages("PopGenome")

library(vcfR)
library(PopGenome)
library(pegas)

# Load the data
UCP1 <- readData("~Desktop/final/se_UCP1.vcf.gz", format="VCF")
```



Analysis:
```{r}
VERgenome <- readData("~/Desktop/final/pyg_UCP1.vcf", format = "VCF")
VERgenome
VER <- read.vcfR("VER/pyg_UCP1.vcf", verbose=T, nrows=284)
VERdna <- vcfR2DNAbin(VER)
VERdna
neut <- neutrality.stats(VERgenome)
get.neutrality(neut)
neut@Fu.Li.F
neut@Fu.Li.D
neut@Tajima.D
tajima <- tajima.test(VERdna)
tajima
```


